src/udev/udevadm-info.c

```diff
@@ -260,6 +265,117 @@ static int print_record(sd_device *device, const char *prefix) {
+                r = json_variant_set_field_string(&v, "R", str);
+                if (r < 0)
+                        return r;
+        }
+
+        if (sd_device_get_subsystem(device, &subsys) >= 0) {
+                r = json_variant_set_field_string(&v, "U", subsys);
+                if (r < 0)
+                        return r;
+        }
+
+        if (sd_device_get_devtype(device, &str) >= 0) {
+                r = json_variant_set_field_string(&v, "T", str);
+                if (r < 0)
+                        return r;
+        }
+
+        if (sd_device_get_devnum(device, &devnum) >= 0) {
+                _cleanup_free_ char *s = NULL;
+
+                if (asprintf(&s,
+                             "%c " DEVNUM_FORMAT_STR,
+                             streq_ptr(subsys, "block") ? 'b' : 'c',
+                             DEVNUM_FORMAT_VAL(devnum)) < 0)
+                        return -ENOMEM;
+
+                r = json_variant_set_field_string(&v, "D", str);
+                if (r < 0)
+                        return r;
+        }
+
+        if (sd_device_get_ifindex(device, &ifi) >= 0) {
+                r = json_variant_set_field_integer(&v, "I", ifi);
+                if (r < 0)
+                        return r;
+        }
+
+        if (sd_device_get_devname(device, &str) >= 0) {
+                _cleanup_(json_variant_unrefp) JsonVariant *a = NULL;
+
+                r = json_variant_set_field_string(&v, "N", str);
+                if (r < 0)
+                        return r;
+
+                if (device_get_devlink_priority(device, &i) >= 0) {
+                        r = json_variant_set_field_integer(&v, "L", i);
+                        if (r < 0)
+                                return r;
+                }
+
+                FOREACH_DEVICE_DEVLINK(device, link) {
+                        const char *val;
+
+                        assert_se(val = path_startswith(link, "/dev/"));
+
+                        r = json_variant_append_arrayb(&a, JSON_BUILD_STRING(val));
+                        if (r < 0)
+                                return r;
+                }
+
+                if (a) {
+                        r = json_variant_set_field(&v, "S", a);
+                        if (r < 0)
+                                return r;
+                }
+        }
+
+        if (sd_device_get_diskseq(device, &q) >= 0) {
+                r = json_variant_set_field_unsigned(&v, "Q", q);
+                if (r < 0)
+                        return r;
+        }
+
+        if (sd_device_get_driver(device, &str) >= 0) {
+                r = json_variant_set_field_string(&v, "V", str);
+                if (r < 0)
+                        return r;
+        }
+
+        FOREACH_DEVICE_PROPERTY(device, key, val) {
+                r = json_variant_set_field_string(&v, key, val);
+                if (r < 0)
+                        return r;
+        }
+
+        *ret = TAKE_PTR(v);
```

Comment:
it's ok to use json_variant_set_field_string() if you must, but also kinda in-effecient. Please use json_build() instead, i.e. query all the bits, and then do:

```c
r = json_build(ret, JSON_BUILD_OBJECT(JSON_BUILD_PAIR_CONDITION(diskseq != UINT64, JSON_BUILD_INTEGER(diskseq)),
JSON_BUILD_PAIR_CONDITION(devnum != 0, JSON_BUILD_OBJECT(JSON_BUILD_PAIR("type", JSON_BUILD_STRING(streq(subsys, "block") ? "block" : "char"), JSON_BUILD_PAIR("major", JSON_BUILD_INTEGER(major(devnum)), …
```

you get the idea